lucene/core/src/java/org/apache/lucene/util/automaton/RegExp.java

```diff
@@ -743,6 +792,30 @@ private Automaton toAutomatonInternal(Map<String,Automaton> automata,
     }
     return a;
   }
+  private Automaton toCaseInsensitiveChar(int codepoint, int maxDeterminizedStates) {
+    Automaton case1 = Automata.makeChar(codepoint);
+    int altCase = Character.isLowerCase(codepoint) ? Character.toUpperCase(codepoint) : Character.toLowerCase(codepoint);
+    Automaton result;
+    if (altCase != codepoint) {
+      result = Operations.union(case1, Automata.makeChar(altCase));
+      result = MinimizationOperations.minimize(result, maxDeterminizedStates);          
+    } else {
+      result = case1;                      
+    }          
+    return result;
+  }
```

Comment:
I think that this is incorrect as there is no 1:1 mapping between lowercase and uppercase letters, for instance `ς` and `σ` both have `Σ` as their uppercase variant. And if someone uses `Σ` in their regexes, `ς` wouldn't match as `toLowerCase(Σ)` returns `σ`.

Should we make this only about ASCII for now, like Java's Pattern class? https://docs.oracle.com/en/java/javase/13/docs/api/java.base/java/util/regex/Pattern.html#CASE_INSENSITIVE We could add support for full Unicode later but this doesn't look like a low hanging fruit to me?

Reply:
good catch, +1 for ASCII only for now, I guess it was too ambitious to handle unicode in the first run

Reply:
Would that be an argument for having a bit mask for case sensitivity flags instead of a Boolean? Java has the UNICODE_CASE flag which differs from the ASCII one and it might be an idea to leave something in the API open to future work.

Reply:
So you mean having two bit mask flags, one for the operators and one for the matching behavior ?  I think that makes sense and could avoid complicating the ctrs in the future so +1.

Reply:
Yes, that's what I was thinking

Reply:
The one downside is RegExpQuery already has a (Term, int, int) constructor for syntax flags and maxDeterminizedStates so I added a (int, int, Term) variation for passing the syntax and match flags.

Reply:
An alternative would be an overhaul of RegExp.
* Introducing a Builder class for the parser with named properties for settings
* separating the RegExp parser logic from the  parsed objects (currently they are the same class). 
* separating rendering functions (toString, to Automaton, toStringTree) from the parsed objects.

I'm not sure if we're at the tipping point where all of that would make sense.

Reply:
>  so I added a (int, int, Term) variation for passing the syntax and match flags.

I am not sure this is clearer ;). It's not ideal but I prefer adding only one ctr `(Term, int, int, int)`.